ast/file.go

```diff
@@ -47,13 +51,31 @@ type FileNode struct {
 // This function panics if the concrete type of any element of decls is not
 // from this package.
 func NewFileNode(info *FileInfo, syntax *SyntaxNode, decls []FileElement, eof Token) *FileNode {
+	return newFileNode(info, syntax, nil, decls, eof)
+}
+
+// NewFileNodeWithEdition creates a new *FileNode. The edition parameter is required. If a file
+// has no edition declaration, use NewFileNode instead.
+//
+// This function panics if the concrete type of any element of decls is not
+// from this package.
+func NewFileNodeWithEdition(info *FileInfo, edition *EditionNode, decls []FileElement, eof Token) *FileNode {
```

Comment:
None of the other constructors in this package use the options patterns since optional items generally aren't combinatoric -- i.e. if we did have another option, it may not necessarily be combinable with the `WithEdition`. So the various concrete functions for a node type effectively enumerate the allowed "shapes" of a given AST node.

I believe the protobuf team is _very_ cautious about changing the grammar/syntax of the language, so it's unlikely that we'll have enough future changes to AST nodes like this to warrant variadic options.